source/Halibut/Transport/ConnectionPool.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Runtime.Serialization;
+using Halibut.Diagnostics;
+using Halibut.Logging;
+
+namespace Halibut.Transport
+{
+    public class ConnectionPool<TKey, TPooledResource>
+        where TPooledResource : class, IPooledResource
+    {
+        readonly ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>> pool = new ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>>();
+
+        public int GetTotalConnectionCount()
+        {
+            return pool.Values.Sum(v => v.Count);
+        }
+
+        public TPooledResource Take(TKey endPoint)
+        {
+            while (true)
+            {
+                var connections = pool.GetOrAdd(endPoint, i => new ConcurrentDictionary<TPooledResource, byte>());
+                AtomicTake(connections, out var connection);
+
+                if (connection == null || !connection.HasExpired())
+                    return connection;
+
+                connection.Dispose();
+            }
+        }
+
+
+
+        public void Return(TKey key, TPooledResource resource)
+        {
+            var resources = pool.GetOrAdd(key, i => new ConcurrentDictionary<TPooledResource, byte>());
+            resources.TryAdd(resource, 0);
+            resource.NotifyUsed();
+
+            while (resources.Count > 5)
+            {
+                if (AtomicTake(resources, out var dispose))
+                {
+                    dispose.Dispose();
+                }
+            }
+        }
+
+        public void Clear(TKey key, Diagnostics.ILog log = null)
+        {
+            if (!pool.TryRemove(key, out var connections))
```

Comment:
What if another thread is on line 25 at this point. We would dispose all the connections and then the other thread would take an already disposed connection out of the pool. We could be halfway though the foreach on ln 56 when this occurs.